### ISSUE TITLE
Merging maps does not work when a value in a map initilizes the "variables" attribute in aws_api_gateway_deployment

### ISSUE THREAD

**Issue by mille-printemps** [hashicorp/terraform#22390]
### Terraform Version

```
Terraform v0.12.6
+ provider.aws v2.14.0
```

### Terraform Configuration Files

```hcl
# foobar/main.tf
resource "aws_api_gateway_rest_api" "bar" {
  name                     = "bar"
  description              = "bar"
  minimum_compression_size = 0

  binary_media_types = [
    "*/*",
    "application/octet-stream",
    "image/*",
    "application/xml",
    "text/xml",
  ]

  endpoint_configuration {
    types = ["REGIONAL"]
  }
}

resource "aws_api_gateway_deployment" "foo" {
  description = "foo"
  rest_api_id = "${aws_api_gateway_rest_api.bar.id}"
  stage_name  = "${var.stage_name}"

  variables = {
    "deployed_at" = "${var.deployed_at}"
  }
}

variable "deployed_at" {}
variable "stage_name" {}

# foo.tf
locals {
  foo = {
    deployed_at = timestamp()
    stage_name = "foo"
  }
}

# bar.tf
locals {
  bar = {
    region = "ap-northeast-1"
  }
}

# locals.tf
locals {
  foobar = merge(local.foo, local.bar)
}

# providers.tf
provider "aws" {
  version = "~> 2.14.0"
  region  = local.foobar["region"]
}

terraform {
  backend "s3" {
    bucket = "some_existing_s3_bucket"
    key    = "foo/terraform.tfstate"
    region = "ap-northeast-1"
  }

  required_version = ">= 0.12.0"
}

# main.tf
module "foobar" {
  source = "./foobar"

  deployed_at = local.foobar["deployed_at"]
  stage_name = local.foobar["stage_name"]
}

```

### Debug Output

```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.


------------------------------------------------------------------------

Error: Invalid AWS Region: 

  on providers.tf line 1, in provider "aws":
   1: provider "aws" {

```

### Expected Behavior

`local.foobar["region"]` is initialized as `ap-northeast-1` and `terraform plan` succeeds. 

### Actual Behavior

It seems that the value of `local.foobar["region"]` is initilized as an empty string and `terraform plan` fails because the region information is not valid. 

### Steps to Reproduce

1. Configure an S3 bucket and set its name to the `backend`
2. Make each terraform configuration files above
3. `terraform init`
4. `terraform plan`

### Additional Context

I come across this issue after I start using version 0.12.x. The terraform configuration files like above were working as expected when I was using version 0.11.x.

Simple merging maps works. As far as I investigated, however, it fails when a value in a map initializes the `variables` attribute in `aws_api_gateway_deployment`, i.e. `deployed_at` in the terraform configuration files above. When I commented out the `deployed_at` variable in files where it exists, `terraform plan` succeeded without any problems.

**Comment by mille-printemps** [comment]
As another additional context, this issue does not occur when the value of `deployed_at` above is a static string, e.g. "foo", but it does when the value is some dynamic string, e.g. "timestamp()" or "random_uuid.foo.result".

**Comment by apparentlymart** [comment]
Hi @mille-printemps!

This issue has been fixed in #24032, which is now merged ready to be included in the forthcoming 0.13.0 release.

**Comment by mille-printemps** [comment]
Thank you all for taking time to fix this.
